# pilot SGE

## HDR library quantification (`nf-sge`)

`nf-sge` is a prototype version of the Nextflow pipeline [QUANTS](https://github.com/cancerit/QUANTS) and was used for HDR library quantification in the pilot SGE experiment. 

Version 0.0.1 of the `nf-sge` pipeline ([nf-sge__version-0.0.1](nf-sge__version-0.0.1)) was run using [Nextflow](https://www.nextflow.io/) version 19.10.0 and Docker (for dependencies e.g. [Cutadapt](https://cutadapt.readthedocs.io/en/stable/)). Included within this repository is the Dockerfile and scripts used to generate that container [nf-sge__docker-version-0.0.5](nf-sge__docker-version-0.0.5). 

To run the pipeline:

```
nextflow [path-to]/nf-sge/main.nf -c experiment.config
```

Below is an example of the `experiment.config` used with `nf-sge` version 0.0.1:

```
docker {
	enabled = true
	runOptions = "-v /home/ubuntu/:/home/ubuntu/"
}

process {
	container = "quay.io/vaofford/nf-sge:0.0.5"
}

params {
	//reads
	reads = [full path to reads]
	library = [full path to library]
	analysis = "library"

	// cutadapt
	cutadapt_pair_adapters = true
	cutadapt_g = 'file:[full path to R1 adapters]'
	cutadapt_G = 'file:[full path to R2 adapters]'

	cutadapt_error_rate = 0.5
	cutadapt_extra_options = false

	// library quantification
	wildtype_sequence = [wildtype sequence]
}
```

Which requires the adapters in [adapters](adapters) for read trimming with cutadapt.

## Indel quantification (`CRISPRESSO2` + bespoke R script)

[CRISPRESSO2](http://crispresso2.pinellolab.org/) version 2.0.34 ([Docker](https://hub.docker.com/layers/pinellolab/crispresso2/v2.0.34/images/sha256-ab5de4bc442ab3b138ebda12b8c52e388a2500383b33eed7a23d428b0bca235b?context=explore)) was used to get allele frequencies (`*..alleles_frequency_table.tsv`). 
Example CRISPRESSO2 command:

```
docker run -v ${PWD}:/DATA -w /DATA pinellolab/crispresso2:v2.0.34 \
	CRISPResso \
		-r1 [path to FASTQ] \
		-a [amplicon sequence] \
		-an [amplicon name] \
		-g [guide] \
		-gn [guide name] \
		-c [coding sequence] \
		-n [experiment name] \
		-o [output prefix] \
		--exclude_bp_from_left 0 \
		--exclude_bp_from_right 0
```

A bespoke R script (aln2cigar.R)[aln2cigar.R] was used to determine indel frequencies using a Docker container with R and package dependencies installed [https://github.com/team113sanger/t113-docker-rbase/releases/tag/2.0.3](https://github.com/team113sanger/t113-docker-rbase/releases/tag/2.0.3). 

Example of command for `aln2cigar.R`:

```
aln2cigar.R \
	--counts [path to count file] \
	--ext '.alleles_frequency_table.tsv' \
	--amplicon [amplicon sequence] \
	--out [path to output directory]
```

For each variant, the scripts aligns reference and alternative sequences in the allele frequency table generated by CRISPRESSO2 and builds the corresponding CIGAR string. Finally, it summarises the frequency of each CIGAR string for the experiment.